### PR TITLE
HDFS-16542. Fix failed unit tests in branch branch-3.2

### PR DESCRIPTION
**Pull request by cndaimin**
Tests fail in branch branch-3.2:
```
hadoop.hdfs.tools.offlineImageViewer.TestOfflineImageViewer
hadoop.hdfs.TestReconstructStripedFileWithValidator
hadoop.hdfs.server.namenode.TestFsck
hadoop.hdfs.server.blockmanagement.TestBlockManager
```

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   8m 15s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 3 new or modified test files.  |
|||| _ branch-3.2 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  30m 41s |  |  branch-3.2 passed  |
| +1 :green_heart: |  compile  |   1m  6s |  |  branch-3.2 passed  |
| +1 :green_heart: |  checkstyle  |   0m 53s |  |  branch-3.2 passed  |
| +1 :green_heart: |  mvnsite  |   1m 11s |  |  branch-3.2 passed  |
| +1 :green_heart: |  javadoc  |   1m  6s |  |  branch-3.2 passed  |
| +1 :green_heart: |  spotbugs  |   3m 16s |  |  branch-3.2 passed  |
| +1 :green_heart: |  shadedclient  |  15m 57s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m  7s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 56s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 56s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   0m 39s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m  3s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 49s |  |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   2m 50s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  16m 38s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  | 185m 22s | [/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4177/1/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt) |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 47s |  |  The patch does not generate ASF License warnings.  |
|  |   | 270m 16s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.datanode.TestDirectoryScanner |
|   | hadoop.hdfs.server.blockmanagement.TestUnderReplicatedBlocks |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4177/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/4177 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell |
| uname | Linux e5ac0aa45b51 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | branch-3.2 / 5a3d0795fd0fc209176cc39ad81e24c6111926e8 |
| Default Java | Private Build-1.8.0_312-8u312-b07-0ubuntu1~18.04-b07 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4177/1/testReport/ |
| Max. process+thread count | 2848 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4177/1/console |
| versions | git=2.17.1 maven=3.6.0 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

**Comment by Hexiaoqiao** [comment]
Thanks @cndaimin for your nice work. Do you mind to check the failed unit tests? Thanks.

**Comment by cndaimin** [comment]
@Hexiaoqiao  OK. Some tests are unstable, updated.

**Review by aajisaka — CHANGES_REQUESTED** [review]
-1

How about backporting the fixes in upstream such as HDFS-13983 instead of fixing multiple unit tests at a time?

File: hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/TestOfflineImageViewer.java [hunk 1]
```diff
@@ -838,7 +838,7 @@ public void testReverseXmlWithoutSnapshotDiffSection() throws Throwable {
       writer.println("<?xml version=\"1.0\"?>");
       writer.println("<fsimage>");
       writer.println("<version>");
-      writer.println("<layoutVersion>-67</layoutVersion>");
+      writer.println("<layoutVersion>-65</layoutVersion>");
```
Comment (aajisaka) [inline hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/TestOfflineImageViewer.java:841]: The change itself is good. Would you create a separate issue and separate PR for this fix?

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 37s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 5 new or modified test files.  |
|||| _ branch-3.2 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  30m 37s |  |  branch-3.2 passed  |
| +1 :green_heart: |  compile  |   1m  2s |  |  branch-3.2 passed  |
| +1 :green_heart: |  checkstyle  |   0m 53s |  |  branch-3.2 passed  |
| +1 :green_heart: |  mvnsite  |   1m 11s |  |  branch-3.2 passed  |
| +1 :green_heart: |  javadoc  |   1m  5s |  |  branch-3.2 passed  |
| +1 :green_heart: |  spotbugs  |   3m  8s |  |  branch-3.2 passed  |
| +1 :green_heart: |  shadedclient  |  16m 14s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m  5s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 58s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 58s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   0m 40s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m  3s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 50s |  |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   2m 50s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  16m 38s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  | 182m 32s | [/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4177/2/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt) |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 48s |  |  The patch does not generate ASF License warnings.  |
|  |   | 259m 38s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.blockmanagement.TestBlockTokenWithShortCircuitRead |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4177/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/4177 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell |
| uname | Linux a529b2e4057a 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | branch-3.2 / 649b54c7250520127bf7887661fd1cfda08ade18 |
| Default Java | Private Build-1.8.0_312-8u312-b07-0ubuntu1~18.04-b07 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4177/2/testReport/ |
| Max. process+thread count | 2889 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4177/2/console |
| versions | git=2.17.1 maven=3.6.0 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

**Comment by cndaimin** [comment]
> -1
> 
> How about backporting the fixes in upstream such as [HDFS-13983](https://issues.apache.org/jira/browse/HDFS-13983) instead of fixing multiple unit tests at a time?

@aajisaka Yes, that's a better choice. I will start another PR to fix the uncovered failed tests and close this one. Please help backporting HDFS-13983. Thanks.

File: hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/TestOfflineImageViewer.java (hunk 1 above)
Comment (aajisaka) [reply hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/TestOfflineImageViewer.java:841]: Fixed by https://issues.apache.org/jira/browse/HDFS-16546. Thanks

**Comment by aajisaka** [comment]
> Please help backporting [HDFS-13983](https://issues.apache.org/jira/browse/HDFS-13983). Thanks.

Done. Thanks!
https://github.com/apache/hadoop/commit/27621c9eef244f6edbb26c099c575e1558efa869
